File: JAVA/650. 2 Keys Keyboard.java
class Solution {
    public int minSteps(int n) {
       if(n==1)
       return 0;
       if(n==2)
       return 2;
       int[] ans = new int[n+1];
       ans[1]=0;
       ans[2]=2;
       for(int i=3;i<=n;i++)
       {
          ans[i]=i;
          int j = i/2;           
           while(j>=1)
           {
             if(i%j==0)
              ans[i] = Math.min(ans[i], ans[j]+i/j);
            j--;
           }
       } 
       return ans[n];
    }
}

File: C++/650. 2 Keys Keyboard.cpp
class Solution {
public:
    int helper(const int& n, int count_A, int paste_A)
    {
        if (count_A == n)
            return 0;
        if (count_A > n)
            return 1000;
        int r1 = 1 + helper(n, count_A + paste_A, paste_A),
            r2 = 2 + helper(n, 2 * count_A, count_A);
        return r1 < r2 ? r1 : r2;
    }
    int minSteps(int n) {
        if (n == 1)
            return 0;
        return 1 + helper(n, 1, 1);
    }
};
